This patch allows dock icons, appicons and miniwindows to be activated with
a single-click (as opposed to a double-click). Great for weary fingers!

Index: WPrefs.app/Expert.c
===================================================================
RCS file: /cvsroot/wm/WPrefs.app/Expert.c,v
retrieving revision 1.25
diff -u -r1.25 Expert.c
--- WPrefs.app/Expert.c	23 Oct 2004 03:30:04 -0000	1.25
+++ WPrefs.app/Expert.c	23 Oct 2004 04:16:07 -0000
@@ -54,6 +54,7 @@
     WMSetButtonSelected(panel->swi[4], GetBoolForKey("DontConfirmKill"));
     WMSetButtonSelected(panel->swi[5], GetBoolForKey("DisableBlinking"));
     WMSetButtonSelected(panel->swi[6], GetBoolForKey("AntialiasedText"));
+    WMSetButtonSelected(panel->swi[7], GetBoolForKey("SingleClickLaunch"));
 }
 
 
@@ -66,7 +67,7 @@
     panel->box = WMCreateBox(panel->parent);
     WMSetViewExpandsToParent(WMWidgetView(panel->box), 2, 2, 2, 2);
 
-    for (i=0; i<7; i++) {
+    for (i=0; i<8; i++) {
         panel->swi[i] = WMCreateSwitchButton(panel->box);
         WMResizeWidget(panel->swi[i], FRAME_WIDTH-40, 25);
         WMMoveWidget(panel->swi[i], 20, 20+i*25);
@@ -79,6 +80,7 @@
     WMSetButtonText(panel->swi[4], _("Disable confirmation panel for the Kill command."));
     WMSetButtonText(panel->swi[5], _("Disable selection animation for selected icons."));
     WMSetButtonText(panel->swi[6], _("Smooth font edges (needs restart)."));
+    WMSetButtonText(panel->swi[7], _("Launch applications and restore windows with a single click."));
 
     WMSetButtonEnabled(panel->swi[6], True);
 
@@ -103,6 +105,7 @@
     SetBoolForKey(WMGetButtonSelected(panel->swi[4]), "DontConfirmKill");
     SetBoolForKey(WMGetButtonSelected(panel->swi[5]), "DisableBlinking");
     SetBoolForKey(WMGetButtonSelected(panel->swi[6]), "AntialiasedText");
+    SetBoolForKey(WMGetButtonSelected(panel->swi[7]), "SingleClickLaunch");
 }
 
 
Index: src/WindowMaker.h
===================================================================
RCS file: /cvsroot/wm/src/WindowMaker.h,v
retrieving revision 1.60
diff -u -r1.60 WindowMaker.h
--- src/WindowMaker.h	18 Oct 2004 01:30:02 -0000	1.60
+++ src/WindowMaker.h	23 Oct 2004 04:16:07 -0000
@@ -472,6 +472,9 @@
     unsigned int workspace_border_size; /* Size in pixels of the workspace border */
     char workspace_border_position;     /* Where to leave a workspace border */
 
+    /* single click to lauch applications */
+    char single_click;
+
     struct {
         unsigned int nodock:1;	       /* don't display the dock */
         unsigned int noclip:1;         /* don't display the clip */
Index: src/appicon.c
===================================================================
RCS file: /cvsroot/wm/src/appicon.c,v
retrieving revision 1.50
diff -u -r1.50 appicon.c
--- src/appicon.c	13 Oct 2004 05:09:08 -0000	1.50
+++ src/appicon.c	23 Oct 2004 04:16:08 -0000
@@ -590,6 +590,7 @@
     Bool movingSingle = False;
     int oldX = x;
     int oldY = y;
+    Bool hasMoved = False;
 
     if (aicon->editing || WCHECK_STATE(WSTATE_MODAL))
         return;
@@ -665,6 +666,7 @@
             break;
 
         case MotionNotify:
+            hasMoved = True;
             if (!grabbed) {
                 if (abs(dx-ev.xmotion.x)>=MOVE_THRESHOLD
                     || abs(dy-ev.xmotion.y)>=MOVE_THRESHOLD) {
@@ -802,6 +804,9 @@
             if (wPreferences.auto_arrange_icons)
                 wArrangeIcons(scr, True);
 
+            if (wPreferences.single_click && !hasMoved)
+                iconDblClick(desc, event);
+
             done = 1;
             break;
         }
Index: src/defaults.c
===================================================================
RCS file: /cvsroot/wm/src/defaults.c,v
retrieving revision 1.90
diff -u -r1.90 defaults.c
--- src/defaults.c	22 Oct 2004 23:39:11 -0000	1.90
+++ src/defaults.c	23 Oct 2004 04:16:08 -0000
@@ -562,6 +562,9 @@
     {"DisableBlinking",	"NO",		NULL,
     &wPreferences.dont_blink,	getBool,	NULL
     },
+    {"SingleClickLaunch",	"NO",	NULL,
+    &wPreferences.single_click,	getBool,	NULL
+    },
     /* style options */
     {"MenuStyle", 	"normal",  		seMenuStyles,
     &wPreferences.menu_style, getEnum, 	setMenuStyle
Index: src/dock.c
===================================================================
RCS file: /cvsroot/wm/src/dock.c,v
retrieving revision 1.91
diff -u -r1.91 dock.c
--- src/dock.c	23 Oct 2004 03:30:04 -0000	1.91
+++ src/dock.c	23 Oct 2004 04:16:09 -0000
@@ -3758,7 +3758,7 @@
 
 
 
-static void
+static Bool
 handleIconMove(WDock *dock, WAppIcon *aicon, XEvent *event)
 {
     WScreen *scr = dock->screen_ptr;
@@ -3776,6 +3776,7 @@
     Bool docked;
     int superfluous = wPreferences.superfluous; /* we catch it to avoid problems */
     int omnipresent = aicon->omnipresent; /* this must be cached!!! */
+    Bool hasMoved = False;
 
 
     if (wPreferences.flags.noupdates)
@@ -3828,6 +3829,7 @@
             break;
 
         case MotionNotify:
+            hasMoved = True;
             if (!grabbed) {
                 if (abs(ofs_x-ev.xmotion.x)>=MOVE_THRESHOLD
                     || abs(ofs_y-ev.xmotion.y)>=MOVE_THRESHOLD) {
@@ -3977,9 +3979,11 @@
 #ifdef DEBUG
             puts("End icon move");
 #endif
-            return;
+            return hasMoved;
         }
     }
+
+    return False;  /* never reached */
 }
 
 
@@ -4116,8 +4120,11 @@
                 handleClipChangeWorkspace(scr, event);
             else
                 handleDockMove(dock, aicon, event);
-        } else
-            handleIconMove(dock, aicon, event);
+        } else {
+            Bool hasMoved = handleIconMove(dock, aicon, event);
+            if (wPreferences.single_click && !hasMoved)
+                iconDblClick(desc, event);
+        }
 
     } else if (event->xbutton.button==Button2 && dock->type==WM_CLIP &&
                aicon==scr->clip_icon) {
Index: src/icon.c
===================================================================
RCS file: /cvsroot/wm/src/icon.c,v
retrieving revision 1.27
diff -u -r1.27 icon.c
--- src/icon.c	16 Oct 2004 22:05:04 -0000	1.27
+++ src/icon.c	23 Oct 2004 04:16:10 -0000
@@ -891,6 +891,7 @@
     int dx=event->xbutton.x, dy=event->xbutton.y;
     int grabbed=0;
     int clickButton=event->xbutton.button;
+    Bool hasMoved = False;
 
     if (WCHECK_STATE(WSTATE_MODAL))
         return;
@@ -942,6 +943,7 @@
             break;
 
         case MotionNotify:
+            hasMoved = True;
             if (!grabbed) {
                 if (abs(dx-ev.xmotion.x)>=MOVE_THRESHOLD
                     || abs(dy-ev.xmotion.y)>=MOVE_THRESHOLD) {
@@ -979,6 +981,8 @@
 
             if (wPreferences.auto_arrange_icons)
                 wArrangeIcons(wwin->screen_ptr, True);
+            if (wPreferences.single_click && !hasMoved)
+                miniwindowDblClick(desc, event);
             return;
 
         }
